onnxruntime/test/logging_apis/test_logging_apis.cc

```diff
@@ -0,0 +1,388 @@
+                                               onnxruntime::logging::LoggingManager::InstanceType::Temporal};
+
+  std::unique_ptr<onnxruntime::logging::Logger> logger = manager.CreateLogger(log_id);
+  Ort::Logger cpp_ort_logger{reinterpret_cast<const OrtLogger*>(logger.get())};
+
+  // Test Ort::Logger::GetLoggingSeverityLevel
+  OrtLoggingLevel logger_severity = cpp_ort_logger.GetLoggingSeverityLevel();
+  ASSERT_EQ(logger_severity, static_cast<OrtLoggingLevel>(min_severity));
+
+  //
+  // Test successful calls to ORT_CXX_LOGF and ORT_CXX_LOGF_NOEXCEPT.
+  //
+
+  // No variadic args
+  line_num = __LINE__ + 1;
+  ORT_CXX_LOGF(cpp_ort_logger, OrtLoggingLevel::ORT_LOGGING_LEVEL_ERROR, "Hello");
+
+  // Two variadic lvalue args
+  line_num = __LINE__ + 1;
+  ORT_CXX_LOGF(cpp_ort_logger, OrtLoggingLevel::ORT_LOGGING_LEVEL_ERROR, "Hello %s %d", log_id, line_num);
+
+  // No variadic args
+  line_num = __LINE__ + 1;
+  ORT_CXX_LOGF_NOEXCEPT(cpp_ort_logger, OrtLoggingLevel::ORT_LOGGING_LEVEL_ERROR, "Hello");
+
+  // Two literal variadic args
+  line_num = __LINE__ + 1;
+  ORT_CXX_LOGF_NOEXCEPT(cpp_ort_logger, OrtLoggingLevel::ORT_LOGGING_LEVEL_ERROR, "Hello %s %f", "world", 1.2f);
+
+  //
+  // The following calls should be filtered out due to insufficient severity.
+  //
+
+  line_num = __LINE__ + 1;
+  ORT_CXX_LOGF(cpp_ort_logger, OrtLoggingLevel::ORT_LOGGING_LEVEL_INFO, "Ignored");
+
+  line_num = __LINE__ + 1;
+  ORT_CXX_LOGF_NOEXCEPT(cpp_ort_logger, OrtLoggingLevel::ORT_LOGGING_LEVEL_INFO, "Ignored %d", line_num);
+
+  //
+  // Test errors due to formatting error.
+  //
+
+  // Catch expected exception from ORT_CXX_LOGF macro.
+  try {
+    line_num = __LINE__ + 1;
+    ORT_CXX_LOGF(cpp_ort_logger, OrtLoggingLevel::ORT_LOGGING_LEVEL_ERROR, "%ls", "abc");
+    FAIL();
+  } catch (const Ort::Exception& excpt) {
+    ASSERT_THAT(excpt.what(), testing::HasSubstr("Failed to log message due to formatting error"));
+  }
+
+  // The formatting error is ignored with the ORT_CXX_LOGF_NOEXCEPT macro
+  line_num = __LINE__ + 1;
+  ORT_CXX_LOGF_NOEXCEPT(cpp_ort_logger, OrtLoggingLevel::ORT_LOGGING_LEVEL_ERROR, "%ls", "abc");
+}
+
+TEST_F(MockCAPITestsFixture, CppLogMacroBypassCApiCall) {
+  // Tests the ability of the ORT_CXX_LOG* macros to completely bypass calling the C logging APIs.
+  // We do this by creating a mock OrtApi struct that is hardcoded to return a specific error when Logger_LogMessage
+  // is called.
+
+  const std::string log_id = "MOCK_API";
+  const onnxruntime::logging::Severity min_severity = onnxruntime::logging::Severity::kWARNING;
+
+  onnxruntime::logging::LoggingManager manager{std::make_unique<MockSink>(), min_severity, false,
+                                               onnxruntime::logging::LoggingManager::InstanceType::Temporal};
+  std::unique_ptr<onnxruntime::logging::Logger> logger = manager.CreateLogger(log_id);
+
+  // Create a mock OrtApi.
+  OrtApi mock_ort_api{};
```

Comment:
Is this controversial? This is the easiest way I can think of to mock the C API.

Reply:
It is fine in my book.